### PR TITLE
Support constraints for intermediate values plot

### PR DESCRIPTION
**Pull request by adjeiv**
## Motivation
References https://github.com/optuna/optuna/issues/4830

## Description of the changes
Support constraints with the intermediate values plots, displaying infeasible trials in #CCCCCC colour.
Tests added accordingly.

```python
import optuna
from optuna.Trial import Trial, FrozenTrial
from optuna.Study import create_study
from optuna.visualization import plot_intermediate_values
from typing import Sequence

def objective_with_constraints(trial: Trial) -> float:
        trial.set_user_attr("constraint", [trial.number % 2])

        trial.report(trial.number, step=0)
        trial.report(trial.number + 1, step=1)
        return 0.0

def constraints(trial: FrozenTrial) -> Sequence[float]:
    return trial.user_attrs["constraint"]

study = create_study(sampler=optuna.samplers.NSGAIIISampler(constraints_func=constraints))
study.optimize(objective_with_constraints, n_trials=10)

plot_intermediate_values(study)
```

Plot is below:
image

### CODE REVIEW

**Comment by adjeiv** [comment]
Re. failing docs build, https://github.com/readthedocs/readthedocs.org/issues/10279 seems relevant... Pinning the sphinx version to be less than 7 should fix this

**Comment by not522** [comment]
Thank you for your PR and the investigation for CI fail. The Optuna maintainers are working to fix the CI. The fix is not necessary to merge this PR, so don't worry about it.

**Comment by not522** [comment]
@Alnusjaponica Could you review this PR?

File: optuna/visualization/_intermediate_values.py
Comment (Alnusjaponica) [inline optuna/visualization/_intermediate_values.py:135]: Using `|` as a union operator for `dict` is only available since `Python 3.9`. Several ways to do it before `Python 3.9` is mentioned in [PEP584](https://peps.python.org/pep-0584/), so please use one of them.

**Review by Alnusjaponica — CHANGES_REQUESTED** [review]
Thanks for your contribution. 
Could you make a similar change in `optuna/visualization/matplotlib/_intermediate_values.py`?

File: tests/visualization_tests/test_intermediate_plot.py
Comment (Alnusjaponica) [inline tests/visualization_tests/test_intermediate_plot.py:103]: By adding following example, you'll see the test fail when `python<=3.8`
ref. https://github.com/optuna/optuna/pull/4851#discussion_r1282631566
```
        _IntermediatePlotInfo(
            trial_infos=[
                _TrialInfo(
                    trial_number=0, sorted_intermediate_values=[(0, 1.0), (1, 2.0)], feasible=True
                ),
                _TrialInfo(
                    trial_number=1, sorted_intermediate_values=[(1, 2.0), (0, 1.0)], feasible=False
                ) 
            ]
        )
```

File: optuna/visualization/_intermediate_values.py
Comment (Alnusjaponica) [inline optuna/visualization/_intermediate_values.py:135]: Sorry for the confusion but `dict.update()` returns `None` like `list.sort()` Please change this part like this.
```suggestion
            else {**default_marker, "color": "#CCCCCC"},
```

**Review by Alnusjaponica — CHANGES_REQUESTED** [review]
Thanks for your prompt response! 
I have added some comments and would appreciate it if you could review them.

File: optuna/visualization/_intermediate_values.py
Comment (adjeiv) [reply optuna/visualization/_intermediate_values.py]: Of course - silly mistake. Thank you :)

**Comment by Alnusjaponica** [comment]
Thanks! LGTM. 
Let me approve this PR after we confirm CI success.

**Review by not522 — COMMENTED** [review]
Thank you! LGTM except the format. Please check the `Checks` CI.

**Review by not522 — APPROVED** [review]
LGTM!

**Review by Alnusjaponica — APPROVED** [review]
Thanks! LGTM.

**Pull request merged**
